### ISSUE TITLE
Can create a resource

### ISSUE THREAD

**Issue by joelmccracken** [noahzgordon/elm-jsonapi#29]
I'm creating this as a ticket related to https://github.com/noahzgordon/elm-jsonapi-http/issues/2

I looked at this last night, and one of the things ran into is that the shape of data for creating a resource with a json api.

From the json api spec doc, a create should look like this:
```
POST /photos HTTP/1.1
Content-Type: application/vnd.api+json
Accept: application/vnd.api+json

{
  "data": {
    "type": "photos",
    "attributes": {
      "title": "Ember Hamster",
      "src": "http://example.com/images/productivity.png"
    },
    "relationships": {
      "photographer": {
        "data": { "type": "people", "id": "9" }
      }
    }
  }
}
```

Some differences
- no links (it seems -- there is def ambiguity in the spec).
- no jsonapi spec version.
- no meta
- no included records
- no "one or many" resources, only one.

There are more, though.

I'm feeling that the JSON api spec is a little unclear/ambiguous in places, especially for creating/updating resources.

**Comment by noahzgordon** [comment]
This makes sense to me:
+ links describe where to find resources linked to the current resource; as a consequence they're a read-only property described by the server and can't be POSTed. The response can specify any resource linkages to the resource you've created.
+ The JSON API spec version can probably be specified in the content-type if future versions of JSON API change the resource structure (Which I don't think they can, since it's a backwards-compatible spec)
+ Meta and included resources also only really make sense as read-only properties provided by the server. The resource is defined by its attributes and relationships to resources that already exist.
+ I'm not sure about posing a resource collection, but I think you would just provide an array of objects to "data" instead of an object. The spec could be clearer here.

**Comment by noahzgordon** [comment]
Actually, I take back that last point.
```
The request MUST include a single resource object as primary data. The resource object MUST contain at least a type member.
```

**Comment by ThomasWeiser** [comment]
Great outline of this feature so far!

For a project I'm currently working on I've [written a function](https://github.com/oliverbarnes/participate/blob/aada1916685d6621f4721fb5c69b0c4d9825ab54/src/JsonApi/Extra.elm#L43-L139) to encode a new resource object from its type, attributes and relationships:

```elm
encodeDocument :
    String
    -> Maybe String
    -> List ( String, Encode.Value )
    -> List ( String, ResourceLinkage )
    -> String
```

In it's current state it's mainly intended for the specific use case at hand. But maybe it's useful also as a first draft for something to be included in this library.

**Comment by noahzgordon** [comment]
@ThomasWeiser I'm taking a look at this idea now. Thanks for the prototype!

The first step to submitting data with `elm-jsonapi-http` will be to get a good encoding pattern into `elm-jsonapi`. I'm thinking that a collection of functions representing a "builder" pattern might be the most convenient to create `Resource` objects from scratch, accompanied by a function to encode any given `Resource` to a JSON value.

```
{- JsonApi.Resource.Builder -}

resource : String -> Resource

withAttributes : Resource -> List ( String, Encode.Value ) -> Resource

withRelationships : Resource -> List ( String, ResourceLinkage ) -> Resource

{- JsonApi.Encode -}

resource : Resource -> Json.Encode.Value
```

And then down the line...

```
{- JsonApi.Http -}

post : String -> Resource -> Request JsonApi.Document

put : Resource -> Request JsonApi.Document
```

The only things missing from this is a nice interface for updating an existing Resource's attributes, which I'm thinking about now.

Let me know what you think about this plan @ThomasWeiser @joelmccracken. I appreciate your feedback :D

**Comment by ThomasWeiser** [comment]
Your approach of using discrete builder functions is certainly more in line with the style of the API so far. Have no strong opinion here.

If you choose the builder route, the arguments should be swapped for application with forward-pipes:

```elm
withAttributes : List ( String, Encode.Value ) -> Resource -> Resource

resource "..."
    |> withAttributes [ ... ]
    |> withRelationships [ ... ]
```

And you may also provide `withAttribute`, `withRelationship` (singular) and also `withId`.

Is there a use case for updating existing resources?

**Comment by joelmccracken** [comment]
Personally, I think having the ability to update existing resources is pretty important.

I like builder-like apis. 

My only real concern with the listed API is using `Resource` for everything. You couldn't take a `Resource` (as defined here https://github.com/noahzgordon/elm-jsonapi/blob/2.0.2/src/JsonApi/Data.elm#L21) that is the result of a pull and directly push it to the server (as discussed above, the json api spec doesn't allow for create submission documents to have resources with relationships that have attributes.). So it seems like it would be good to make `Resource` parameterized with a type: `Resource Fetch`, `Resource Create`, `Resource Update`, `Resource Delete`, etc.

**Comment by noahzgordon** [comment]
@joelmccracken as I understand it, Resource Objects as defined by the JSON API spec never have relationship objects with attributes; those related attributes are pulled from the side-loaded `included` Resources.

Right now I'm toying around with having a `ClientResource` type to represent Resources built on the client without `id`s, as opposed to the `Resource` type which represents Resources pulled from the server. You would use `ClientResource` with POST requests and `Resource` with PUT/PATCH/DELETE.

**Comment by joelmccracken** [comment]
So, `Resource` has a `List RawResource` which has `ResourceObject` which has `attributes`. That's what I'm talking about, the type currently allows it.

Resource could also be:

```
type Resource = FetchResource (all the stuff from before) | CreateResource ... | UpdateResource...
```

You could also just use Resource everywhere and make it smart enough to pull things from where you expect them. That would surprise me personally, but tradeoffs. I'm mostly just providing some ideas =)

**Comment by noahzgordon** [comment]
@joelmccracken right, those are implementation details which allow the Resource to reference the included resources of the document it belongs to. The JSON serialization (which must be compliant with the spec) would not include those.

That said, I do think that a resource built for creation is fundamentally different from one retrieved from the server (to which you might apply PUT, PATCH, DELETE) because of its lack of an `id` member. So it makes sense at least to differentiate that one.

I do appreciate the ideas, keep them coming!

**Comment by joelmccracken** [comment]
One thing: the Json API spec says creates may have an ID field, for e.g. client-generated GUIDs. 

"A server MAY accept a client-generated ID along with a request to create a resource. An ID MUST be specified with an id key, the value of which MUST be a universally unique identifier. The client SHOULD use a properly generated and formatted UUID as described in RFC 4122 [RFC4122]."

**Comment by noahzgordon** [comment]
@joelmccracken @ThomasWeiser hey folks, just pushed a new version. It includes client resource encoding, which will hopefully make things easier for you. Check it out and let me know what you think: http://package.elm-lang.org/packages/noahzgordon/elm-jsonapi/2.2.0/JsonApi

Next steps: `POST` requests from `elm-jsonapi-http` taking advantage of this new encoding feature, and the ability to update existing resources to enable `PUT` and `PATCH`.

**Comment by ThomasWeiser** [comment]
Hey @noahzgordon, great progress, thanks!

My thoughts so far:

* When adding a relationship one should be able to construct `ResourceIdentifier` from `type` and `id`, because that may be the way the app models the data. So I think there should be a public function with type `Sting -> String -> ResourceIdentifier`.
* Documentation of `JsonApi-Resources.elm` can be more clear if the functions are arranged in some meaningful groups [here](https://github.com/noahzgordon/elm-jsonapi/blob/1b5021df166ed29d388f79ba8528aeb3d573e82d/src/JsonApi/Resources.elm#L21-L23).
* About the naming of the functions `withRelationship(s)`:
    * The specs speaks about _to-one_ and _to-many_ relationships. Currently, the function `withRelationships` adds _one_ relationship _to many_ resources. IMHO the use of plural is confusing here.
    * So I would propose something like these as function names: `withRelationshipToOne`,  `withRelationshipToMany`
* The specs speaks clearly about _empty top-one_ relationships. So the type of `withRelationship(ToOne)` should reflect this IMHO: `String -> Maybe ResourceIdentifier -> ClientResource -> ClientResource`. That's the same reasoning as [here](https://github.com/noahzgordon/elm-jsonapi/issues/28#issuecomment-263081525) regarding `relatedResource`.

**Comment by noahzgordon** [comment]
@ThomasWeiser all great ideas! Will consider them for the next design session.
